### PR TITLE
improve compliance

### PR DESCRIPTION
**Pull request by rafikurnia**
**CHANGES**
- Remove currently unneeded resources 
- Add Default Encryption to All S3 Buckets
- Enable Server Side Encryption to All S3 Buckets
- Enable Secure Transport to All S3 Buckets
- Enable Versioning on All S3 Buckets

**IMPORTANT NOTES**
Before using this module version, it is mandatory to destroy SG resources for CodeBuild. 
If not then what will be shown on `terraform plan` is:
```bash
Error: Error asking for user input: 1 error(s) occurred:

* module.shared_resources.module.codebuild_sg_name.random_id.this: configuration for module.shared_resources.module.codebuild_sg_name.provider.random is not present; a provider configuration block is required for all operations
```
The command to destroy SG resources for CodeBuild is: 
```bash
terraform destroy -target=module.shared_resources.module.codebuild_sg_name
```
The output of the command above is:
```bash
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  - module.shared_resources.aws_security_group.codebuild

  - module.shared_resources.aws_security_group_rule.codebuild_http_all

  - module.shared_resources.aws_security_group_rule.codebuild_https_all

  - module.shared_resources.aws_security_group_rule.codebuild_template_ssh

  - module.shared_resources.module.codebuild_sg_name.random_id.this


Plan: 0 to add, 0 to change, 5 to destroy.
```

If you destroy them, then what shown on command `terraform plan` will be:
```bash
------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place
  - destroy

Terraform will perform the following actions:

  - module.shared_resources.aws_iam_role_policy.codebuild_policy_ec2

  ~ module.shared_resources.aws_s3_bucket.application_binary
      policy:                                                                                                "{\"Statement\":[{\"Action\":\"s3:PutObject\",\"Condition\":{\"StringEquals\":{\"s3:x-amz-acl\":\"bucket-owner-read\"}},\"Effect\":\"Allow\",\"Principal\":{\"AWS\":\"arn:aws:iam::517530806209:user/jenkins\"},\"Resource\":\"arn:aws:s3:::tsi-appbin-015110552125-e075f32944df2a4f/tsi*\",\"Sid\":\"AppbinWrite\"}],\"Version\":\"2012-10-17\"}" => "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Sid\": \"AppbinWrite\",\n      \"Effect\": \"Allow\",\n      \"Action\": \"s3:PutObject\",\n      \"Resource\": \"arn:aws:s3:::tsi-appbin-015110552125-e075f32944df2a4f/tsi*\",\n      \"Principal\": {\n        \"AWS\": \"arn:aws:iam::517530806209:user/jenkins\"\n      },\n      \"Condition\": {\n        \"Bool\": {\n          \"aws:SecureTransport\": \"true\"\n        },\n        \"StringEquals\": {\n          \"s3:x-amz-acl\": \"bucket-owner-full-control\"\n        }\n      }\n    },\n    {\n      \"Sid\": \"DenyAllUnEncryptedHTTPAccess\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:*\",\n      \"Resource\": \"arn:aws:s3:::tsi-appbin-015110552125-e075f32944df2a4f/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"Bool\": {\n          \"aws:SecureTransport\": \"false\"\n        }\n      }\n    },\n    {\n      \"Sid\": \"DenyIncorrectEncryptionHeader\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:PutObject\",\n      \"Resource\": \"arn:aws:s3:::tsi-appbin-015110552125-e075f32944df2a4f/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"StringNotEquals\": {\n          \"s3:x-amz-server-side-encryption\": \"AES256\"\n        }\n      }\n    },\n    {\n      \"Sid\": \"DenyUnEncryptedObjectUploads\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:PutObject\",\n      \"Resource\": \"arn:aws:s3:::tsi-appbin-015110552125-e075f32944df2a4f/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"Null\": {\n          \"s3:x-amz-server-side-encryption\": \"true\"\n        }\n      }\n    }\n  ]\n}"
      server_side_encryption_configuration.#:                                                                "0" => "1"
      server_side_encryption_configuration.0.rule.#:                                                         "0" => "1"
      server_side_encryption_configuration.0.rule.0.apply_server_side_encryption_by_default.#:               "0" => "1"
      server_side_encryption_configuration.0.rule.0.apply_server_side_encryption_by_default.0.sse_algorithm: "" => "AES256"
      tags.Description:                                                                                      "CodePipeline artifact bucket for tsi services" => "Application Binary bucket for tsi services"
      tags.Name:                                                                                             "tsi-codepipeline-015110552125-de5aad84ceb115d2" => "tsi-appbin-015110552125-e075f32944df2a4f"

  ~ module.shared_resources.aws_s3_bucket.codebuild_cache
      policy:                                                                                                "" => "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Sid\": \"DenyAllUnEncryptedHTTPAccess\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:*\",\n      \"Resource\": \"arn:aws:s3:::tsi-codebuild-cache-015110552125-73f85ac977d9bbe8/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"Bool\": {\n          \"aws:SecureTransport\": \"false\"\n        }\n      }\n    },\n    {\n      \"Sid\": \"DenyIncorrectEncryptionHeader\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:PutObject\",\n \"Resource\": \"arn:aws:s3:::tsi-codebuild-cache-015110552125-73f85ac977d9bbe8/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"StringNotEquals\": {\n          \"s3:x-amz-server-side-encryption\": [\n            \"aws:kms\",\n            \"AES256\"\n          ]\n        }\n      }\n    },\n    {\n      \"Sid\": \"DenyUnEncryptedObjectUploads\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:PutObject\",\n      \"Resource\": \"arn:aws:s3:::tsi-codebuild-cache-015110552125-73f85ac977d9bbe8/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"Null\": {\n          \"s3:x-amz-server-side-encryption\": \"true\"\n        }\n      }\n    }\n  ]\n}"
      server_side_encryption_configuration.#:                                                                "0" => "1"
      server_side_encryption_configuration.0.rule.#:                                                         "0" => "1"
      server_side_encryption_configuration.0.rule.0.apply_server_side_encryption_by_default.#:               "0" => "1"
      server_side_encryption_configuration.0.rule.0.apply_server_side_encryption_by_default.0.sse_algorithm: "" => "AES256"
      tags.Description:                                                                                      "CodePipeline artifact bucket for tsi services" => "CodeBuild cache bucket for tsi services"
      tags.Name:                                                                                             "tsi-codepipeline-015110552125-de5aad84ceb115d2" => "tsi-codebuild-cache-015110552125-73f85ac977d9bbe8"
      versioning.0.enabled:                                                                                  "false" => "true"

  ~ module.shared_resources.aws_s3_bucket.codepipeline_artifact
      policy:                                                                                                "" => "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Sid\": \"DenyAllUnEncryptedHTTPAccess\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:*\",\n      \"Resource\": \"arn:aws:s3:::tsi-codepipeline-015110552125-de5aad84ceb115d2/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"Bool\": {\n          \"aws:SecureTransport\": \"false\"\n        }\n      }\n    },\n    {\n      \"Sid\": \"DenyIncorrectEncryptionHeader\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:PutObject\",\n      \"Resource\": \"arn:aws:s3:::tsi-codepipeline-015110552125-de5aad84ceb115d2/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"StringNotEquals\": {\n          \"s3:x-amz-server-side-encryption\": [\n            \"aws:kms\",\n            \"AES256\"\n          ]\n        }\n      }\n    },\n    {\n      \"Sid\": \"DenyUnEncryptedObjectUploads\",\n      \"Effect\": \"Deny\",\n      \"Action\": \"s3:PutObject\",\n      \"Resource\": \"arn:aws:s3:::tsi-codepipeline-015110552125-de5aad84ceb115d2/*\",\n      \"Principal\": \"*\",\n      \"Condition\": {\n        \"Null\": {\n          \"s3:x-amz-server-side-encryption\": \"true\"\n  }\n      }\n    }\n  ]\n}"
      server_side_encryption_configuration.#:                                                                "0" => "1"
      server_side_encryption_configuration.0.rule.#:                                                         "0" => "1"
      server_side_encryption_configuration.0.rule.0.apply_server_side_encryption_by_default.#:               "0" => "1"
      server_side_encryption_configuration.0.rule.0.apply_server_side_encryption_by_default.0.sse_algorithm: "" => "AES256"
      versioning.0.enabled:                                                                                  "false" => "true"

  ~ module.shared_resources.aws_security_group_rule.template_codebuild_ssh
      description:                                                                                           "" => "Allow ingress from CodeBuild IP port SSH"

  ~ module.shared_resources.aws_security_group_rule.template_http_all
      description:                                                                                           "" => "Allow egress to all port HTTP"

  ~ module.shared_resources.aws_security_group_rule.template_https_all
      description:                                                                                           "" => "Allow egress to all port HTTPS"


Plan: 0 to add, 6 to change, 1 to destroy.

------------------------------------------------------------------------
```

### CODE REVIEW

File: data.tf [hunk 1]
```diff
@@ -528,6 +529,292 @@ data "aws_iam_policy_document" "appbin_bucket_policy" {
...
+    actions = [
+      "s3:PutObject",
+    ]
+
+    resources = [
+      "arn:aws:s3:::${module.application_binary.name}/*",
+    ]
+
+    condition = {
+      test     = "Null"
```
[the remark this answers is not in the document]
Comment (rafikurnia) [reply data.tf]: > Use a Null condition operator to check if a condition key is present at the time of authorization. In the policy statement, use either true (the key doesn't exist — it is null) or false (the key exists and its value is not null).`

Ref: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html#Conditions_Null

File: data.tf [hunk 2]
```diff
@@ -528,6 +529,292 @@ data "aws_iam_policy_document" "appbin_bucket_policy" {
...
+
+    actions = [
+      "s3:PutObject",
+    ]
+
+    resources = [
+      "arn:aws:s3:::${module.codepipeline_artifact_bucket_name.name}/*",
+    ]
+
+    condition = {
```
[the remark this answers is not in the document]
Comment (rafikurnia) [reply data.tf]: Yes I thought so. Will create CMK for codepipeline per product domain

File: data.tf [hunk 3]
```diff
@@ -528,6 +529,292 @@ data "aws_iam_policy_document" "appbin_bucket_policy" {
         "bucket-owner-read",
```
[the remark this answers is not in the document]
Comment (rafikurnia) [reply data.tf]: got it

[the remark this answers is not in the document]
Comment (salvianreynaldi) [reply data.tf]: this block works, btw, I've been able to successfully upload to the application binary bucket with SSE

File: data.tf [hunk 4]
```diff
@@ -528,6 +529,292 @@ data "aws_iam_policy_document" "appbin_bucket_policy" {
...
+        "*",
+      ]
+    }
+
+    actions = [
+      "s3:*",
+    ]
+
+    resources = [
+      "arn:aws:s3:::${module.codebuild_cache.name}",
```
[the remark this answers is not in the document]
Comment (rafikurnia) [reply data.tf]: at [this page](https://aws.amazon.com/blogs/devops/how-to-enable-caching-for-aws-codebuild/) at the bottom on ***Best practices for cache*** section, the first point says: 
> By default, the cache archive is encrypted on the server side with the customer’s artifact KMS key.

I will try to change the value from `AES256` to `aws:kms`

**Pull request merged**